cmd/cli/plugin/cluster/osimage_oracle.go

```diff
@@ -0,0 +1,189 @@
+	Run: ociPopulateCmdInitRun,
+}
+
+func init() {
+	oraclePopulate.PersistentFlags().StringVar(&compartmentId, "compartment", "", "The destination OCI compartment to import the public image into")
+
+	oracleCmd.AddCommand(oraclePopulate)
+}
+
+// writeToFile writes a Kubernetes runtime.Object to the file, given its paths
+func writeToFile(object runtime.Object, path string) error {
+	serializer := k8sjson.NewSerializerWithOptions(
+		k8sjson.DefaultMetaFactory, nil, nil,
+		k8sjson.SerializerOptions{
+			Yaml:   true,
+			Pretty: true,
+			Strict: true,
+		},
+	)
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	if err = serializer.Encode(object, file); err != nil {
+		return err
+	}
+	return nil
+}
+
+// getTKRFromManifest reads the TanzuKubernetesRelease from its yaml manifest given path
+func getTKRFromManifest(path string) (*v1alpha3.TanzuKubernetesRelease, error) {
+	var tkr v1alpha3.TanzuKubernetesRelease
+	tkrFile, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	if err = utilyaml.NewYAMLOrJSONDecoder(tkrFile, 100).Decode(&tkr); err != nil {
+		return nil, err
+	}
+	return &tkr, nil
+}
+
+// getTKRPath returns the absolute path for the TKR manifest
+func getTKRPath() string {
+	// default TKR manifest file name in the bundle
+	tkrFileName := "TanzuKubernetesRelease.yml"
+	return filepath.Join(outputDirectory, "config", tkrFileName)
+}
+
+// getOSImage returns the absolute path for the OSImage manifest to create
+func getOSImage() string {
+	return filepath.Join(outputDirectory, "config", fmt.Sprintf("%s.yaml", name))
+}
+
+// getOSImageName calculates the default OSImage name, given provided flags
+func getOSImageName(tkrK8sVersion string) string {
+	return fmt.Sprintf("%s-%s-%s-%s-%s", osName, osType, osVersion, osArch, tkrK8sVersion)
+}
+
+// patchTKrWithOSImage patches the TKR bundle with the imported OSImage information
+func patchTKrWithOSImage(tkr *v1alpha3.TanzuKubernetesRelease, imageId, compartment, region string) (err error) {
+	tkr.Spec.OSImages = append(tkr.Spec.OSImages, v1.LocalObjectReference{Name: name})
+	osImage := &v1alpha3.OSImage{
+		TypeMeta: metav1.TypeMeta{Kind: "OSImage", APIVersion: v1alpha3.GroupVersion.Version},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: constants.TkgNamespace,
+		},
+		Spec: v1alpha3.OSImageSpec{
+			KubernetesVersion: tkr.Spec.Kubernetes.Version,
+			OS: v1alpha3.OSInfo{
+				Type:    osType,
+				Name:    osName,
+				Version: osVersion,
+				Arch:    osArch,
+			},
+			Image: v1alpha3.MachineImageInfo{
+				Type: "oci",
+				Ref: map[string]interface{}{
+					"id":          &imageId,
+					"compartment": &compartment,
+					"region":      region,
+				}},
+		},
+	}
+	if err = writeToFile(tkr, getTKRPath()); err != nil {
+		return err
+	}
+	if err = writeToFile(osImage, getOSImage()); err != nil {
```

Comment:
If this fails, should we clean up the tkr written in the previous block?

Reply:
I think it's ok to leave it so the user can at least manually do something with it as an escape hatch.